examples/miral-shell/spinner/miregl.cpp

```diff
@@ -0,0 +1,441 @@
+
+    void make_current(EGLSurface eglsurface) const;
+
+    void swap_buffers(EGLSurface eglsurface) const;
+
+    void destroy_surface(EGLSurface eglsurface) const;
+
+    void get_surface_size(EGLSurface eglsurface, int* width, int* height) const;
+
+    void set_swap_interval(EGLSurface eglsurface, int interval) const;
+
+    bool supports_surfaceless_context();
+
+    ~MirEglApp();
+
+    struct OutputInfo
+    {
+        struct wl_output* wl_output;
+        int32_t x;
+        int32_t y;
+        int32_t width;
+        int32_t height;
+    };
+
+    struct wl_display* const display;
+    struct wl_compositor* compositor;
+    struct wl_shm* shm;
+    struct wl_seat* seat;
+    std::vector<OutputInfo> output_info;
+    struct wl_shell* shell;
+
+private:
+    friend void new_global(
+        void* data,
+        struct wl_registry* registry,
+        uint32_t id,
+        char const* interface,
+        uint32_t version);
+
+    EGLDisplay egldisplay;
+    EGLContext eglctx;
+    EGLConfig eglconfig;
+    EGLint neglconfigs;
+    EGLSurface dummy_surface;
+};
+
+static void output_geometry(void *data,
+    struct wl_output *wl_output,
+    int32_t x,
+    int32_t y,
+    int32_t /*physical_width*/,
+    int32_t /*physical_height*/,
+    int32_t /*subpixel*/,
+    const char */*make*/,
+    const char */*model*/,
+    int32_t /*transform*/)
+{
+    struct MirEglApp* app = static_cast<decltype(app)>(data);
+
+    for (auto& oi : app->output_info)
+    {
+        if (wl_output == oi.wl_output)
+        {
+            oi.x = x;
+            oi.y = y;
+            return;
+        }
+    }
+    app->output_info.push_back({wl_output, x, y, 0, 0});
+}
+
+static void output_mode(void *data,
+    struct wl_output *wl_output,
+    uint32_t flags,
+    int32_t width,
+    int32_t height,
+    int32_t /*refresh*/)
+{
+    if (!(WL_OUTPUT_MODE_CURRENT & flags))
+        return;
+
+    struct MirEglApp* app = static_cast<decltype(app)>(data);
+
+    for (auto& oi : app->output_info)
+    {
+        if (wl_output == oi.wl_output)
+        {
+            oi.width = width;
+            oi.height = height;
+            return;
+        }
+    }
+    app->output_info.push_back({wl_output, 0, 0, width, height});
+}
+
+static void new_global(
+    void* data,
+    struct wl_registry* registry,
+    uint32_t id,
+    char const* interface,
+    uint32_t version)
+{
+    (void)version;
+    struct MirEglApp* app = static_cast<decltype(app)>(data);
+
+    if (strcmp(interface, "wl_compositor") == 0)
+    {
+        app->compositor = static_cast<decltype(app->compositor)>(wl_registry_bind(registry, id, &wl_compositor_interface, 3));
+    }
+    else if (strcmp(interface, "wl_shm") == 0)
+    {
+        app->shm = static_cast<decltype(app->shm)>(wl_registry_bind(registry, id, &wl_shm_interface, 1));
+        // Normally we'd add a listener to pick up the supported formats here
+        // As luck would have it, I know that argb8888 is the only format we support :)
```

Comment:
for now?

Reply:
That's C&P from examples/mirclient/wayland_client.c